Create a symlink to cores directory based on core_pattern

bhavin192 opened the pull request:
Modifies the postStart hook to create a symlink to /mnt/disk0/cores at
the directory being used by kernel.core_pattern. This is useful when
the core_pattern's value points to some absolute path like
/var/lib/cores instead of a relative path and the path doesn't exist
inside the container.

Changing to bash as some of the bash constructs are being used. The
current container image's sh is just a symlink to bash.

Ref: https://github.com/yugabyte/yugabyte-db/issues/4385#issuecomment-717656848

### Scenarios tested
- Installed the chart on nodes having `kernel.core_pattern=/var/lib/test/cores123/core.%e.%p.%t`.
- Exec into the pod,
  ```console
  $ ls -ll /var/lib/test
  total 0
  lrwxrwxrwx 1 root root 16 Nov  2 13:45 cores123 -> /mnt/disk0/cores

  $ /home/yugabyte/dumpit 
  Hello World!
  Segmentation fault (core dumped)

  $ pwd
  /mnt/disk0/cores

  $ ls
  core.dumpit.78.1604324807
  ```

bmatican commented on stable/yugabyte/templates/service.yaml:
```diff
@@ -179,10 +179,33 @@ spec:
+                - "-c"
+                - >
+                  mkdir -p /mnt/disk0/cores;
+                  mkdir -p /mnt/disk0/yb-data/scripts;
```
Should `/mnt/disk0/` be some generic var that a user could override?

Arnav15 replied:
We can make it a generic var, but the dockerfile does hard-code the volumes, so maybe it won't be a good idea.

bmatican commented on stable/yugabyte/templates/service.yaml:
```diff
@@ -179,10 +179,33 @@ spec:
+          - -c
+          - |
+            sysctl -w vm.swappiness=0
+            sysctl -w kernel.core_pattern="/home/yugabyte/cores/core_%e.%p"
```
This symlink is generated by our docker image, right? Is it hardcoded right now to `/mnt/disk0`? Should we switch to having that be generated in the `postStart`, so we could make it a var?

Arnav15 replied:
Sure, but we'll need to decide on how we would want to move forward with the corresponding changes to the dockerfile and what it would mean for the previous docker releases.

bmatican commented on stable/yugabyte/templates/service.yaml:
```diff
@@ -179,10 +179,33 @@ spec:
+                  mkdir -p /mnt/disk0/cores;
+                  mkdir -p /mnt/disk0/yb-data/scripts;
+                  if [ -f /home/yugabyte/bin/log_cleanup.sh ]; then
+                    cp /home/yugabyte/bin/log_cleanup.sh /mnt/disk0/yb-data/scripts;
```
ditto on vars for paths

bmatican commented on stable/yugabyte/templates/service.yaml:
```diff
@@ -276,6 +299,24 @@ spec:
+            done
+        volumeMounts:
+          - name: datadir0
+            mountPath: /home/yugabyte/
```
ah nice, so this now matches the default home var value from the cleanup script, right?

Arnav15 replied:
Yes, it now mimics the home directory path for all the other deployments.

bmatican commented on stable/yugabyte/Chart.yaml:
```diff
@@ -1,7 +1,7 @@
-version: 2.0.2
-appVersion: 2.0.2.0-b10
+version: 2.0.4
+appVersion: 2.0.4.0-b7
```
do we need to include this with the change?

bmatican commented on stable/yugabyte/app-readme.md:
```diff
@@ -1 +1 @@
-This chart bootstraps an RF3 Yugabyte DB version 2.0.2.0-b10 cluster using the Helm Package Manager.
+This chart bootstraps an RF3 Yugabyte DB version 2.0.4.0-b7 cluster using the Helm Package Manager.
```
this feels cumbersome to always bump the version in text here too...

bhavin192 committed:
Create a symlink to cores directory based on core_pattern

Modifies the postStart hook to create a symlink to /mnt/disk0/cores at
the directory being used by kernel.core_pattern. This is useful when
the core_pattern's value points to some absolute path like
/var/lib/cores instead of a relative path and the path doesn't exist
inside the container.

Changing to bash as some of the bash constructs are being used. The
current container image's sh is just a symlink to bash.

Scenarios tested:
- Installed the chart on nodes having `kernel.core_pattern=/var/lib/test/cores123/core.%e.%p.%t`.
- Exec into the pod,
  ```console
  $ ls -ll /var/lib/test
  total 0
  lrwxrwxrwx 1 root root 16 Nov  2 13:45 cores123 -> /mnt/disk0/cores

  $ /home/yugabyte/dumpit
  Hello World!
  Segmentation fault (core dumped)

  $ pwd
  /mnt/disk0/cores

  $ ls
  core.dumpit.78.1604324807
  ```
- Installed the chart where core_pattern is set to pipe to some binary
  with `|/usr/lib/systemd/systemd-coredump %P %u %g %s %t %c %h %e`,
  as expected we don't do anything for such cases (can't to anything).

Ref: https://github.com/yugabyte/yugabyte-db/issues/4385#issuecomment-717656848

Signed-off-by: Bhavin Gandhi <bhavin7392@gmail.com>

ajcaldera1 reviewed (approved):
LGTM.

iSignal commented on stable/yugabyte/templates/service.yaml:
```diff
@@ -218,10 +218,19 @@ spec:
+                    if [ ! -d "${core_dir}" ]; then
+                      # create the parent directory
+                      mkdir -p "${core_dir%/*}";
+                      ln -s "/mnt/disk0/cores" "${core_dir}";
```
we should make sure none of these can cause the postStart to fail, because that would cause the container to never reach running state. Since this step is non-critical, I would add a `|| true` to it

bhavin192 replied:
Yep, made the changes.

iSignal commented on stable/yugabyte/templates/service.yaml:
```diff
@@ -218,10 +218,19 @@ spec:
                 - "-c"
                 - >
                   mkdir -p /mnt/disk0/cores;
+                  pattern="$(sysctl kernel.core_pattern --values || true)";
```
if the sysctl fails, pattern is an empty string right, is that fine?

bhavin192 replied:
Yes, that's fine. It won't go to the inside block where mkdir and ln is run.
